### PR TITLE
Paper: RoughPy

### PR DESCRIPTION
**Pull request by inakleinbottle**
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_:  Charles Lindsey @cdlindsey 

_Reviewers_:  
- Paul Wright @PaulJWright
- Manolis Papadakis @manopapad

### CODE REVIEW

File: papers/sam_morley/roughpy.md [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+records, radio (and other electromagnetic spectra), sound (and speech), etc.
+Traditionally, these data are tricky to work with because of the exponential
+complexity and different scales of the underlying process.
+Until recently, with the development of transformers and large language models,
+it has been difficult to capture the long-term pattern whist also capturing the
+short-term fine detail.
+Rough path theory gives us tools to work with sequential, ordered data in a
+mathematically rigorous way, which should provide a means to overcome some of
+the inherent complexity of the data.
+In this paper, we introduce a new package *RoughPy* for working with sequential
```
Comment (PaulJWright) [inline papers/sam_morley/roughpy.md:43]: While later in the manuscript you refer to DataSig:

> DataSig website (https://datasig.ac.uk/examples).

I think you may be missing a link to the RoughPy Github page or website in this manuscript

File: papers/sam_morley/roughpy.md [hunk 2]
```diff
@@ -0,0 +1,703 @@
...
+  Manuscript version arising from this submission.
+---
+
+## Introduction
+Sequential data appears everywhere in the modern world: text, finance, health 
+records, radio (and other electromagnetic spectra), sound (and speech), etc.
+Traditionally, these data are tricky to work with because of the exponential
+complexity and different scales of the underlying process.
+Until recently, with the development of transformers and large language models,
+it has been difficult to capture the long-term pattern whist also capturing the
```
Comment (PaulJWright) [inline papers/sam_morley/roughpy.md]: typo of "whilst"?

File: papers/sam_morley/roughpy.md [hunk 3]
```diff
@@ -0,0 +1,703 @@
...
+We will construct a stream in $\mathbb{R}^{26}$ by taking each letter in a word,
+"scipy" in this example, as the increments of a path:
+```python
+import numpy as np
+
+text = "scipy"
+increments = np.zeros((5, 26), dtype="int8")
+for i, c in enumerate(text):
+    increments[i, ord(c) - 97] = 1
+```
```
Comment (PaulJWright) [inline papers/sam_morley/roughpy.md]: It could be useful for someone glancing through the manuscript to know that it is pip installable!

**Review by PaulJWright — APPROVED** [review]
Firstly I would like to apologise to the authors as the subject of this paper is outwith my wheelhouse, and as such have provided minor comments where I believe this will help the manuscript.

**Review**: 

This manuscript describes *RoughPy*, a Python package for viewing streams of data through the lens of rough paths and using the tools and techniques of signatures. The manuscript is well-written and clear, with good accompanying documentation at https://roughpy.org/. 

The codebase at https://github.com/datasig-ac-uk/RoughPy seems complete with well-written documentation, testing (including automated testing through Github Actions), and regular releases. There are currently no outstanding issues (https://github.com/datasig-ac-uk/RoughPy/issues) that cause concern, however, it would be nice to merge in https://github.com/datasig-ac-uk/RoughPy/pull/91 or an equivalent. The inclusion of an issue template is noted and appreciated.

I am personally unsure of the reasoning behind the `tests/build (windows-2019)` failure in the CI:

```
FAILED tests/streams/test_lie_increment_path.py::test_from_array_width_3_2_increments - AssertionError: { 1() 3/3(1) 7(2) 1(3) } != { 1() 3(1) 7(2) 1(3) }
```

while `ubuntu-latest`/`macos-11` pass, but would encourage the authors to investigate a fix to this issue.

File: papers/sam_morley/roughpy.md (hunk 3 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md]: Thanks so much for your comments. I completely forgot to merge the code of conduct, thanks for reminding me.

The failed CI on Windows is a very frustrating bug in the Windows compilation of the GNU Multiprecision library, and it is a royal pain that comes and goes seemingly at random. I don't yet have a reason why it fails.

I shall fix the issues that you have highlighted.

File: papers/sam_morley/roughpy.md [hunk 4]
```diff
@@ -0,0 +1,703 @@
...
+functions, including half-shuffle products for `FreeTensor`/`ShuffleTensor`
+objects, and adjoint operators for left free tensor multiplication.
+These are operations that are frequently used in the theory of rough paths, and
+will likely be necessary in developing new applications later (as in the
+signature kernels).
+
+RoughPy algebras are designed around a flexible scalar ring system that allows
+users to perform calculations with different accuracy, or derive expressions by
+using polynomial coefficients.
+For most applications, floating point numbers will provide a good balance
```
Comment (cdlindsey) [inline papers/sam_morley/roughpy.md]: Do you mean 32 bit or 64 bit?  It's worth specifying.  For example, TensorFlow and NumPy have different defaults.

**Comment by cdlindsey** [comment]
Very interesting paper and package.  Could you provide a link to more detailed examples if they are too complex to fully fit into the text?

**Review by cliffckerr — COMMENTED** [review]
I am not a mathematician so I confess most of this paper went over my head, but from what I could tell it was well written. For me it would be helpful if there were more of a link between the very high-level applications (e.g. detecting interference) and the very simple example (computing the signature of the word 'scipy'). I know the authors point to the documentation for more information, but perhaps a slightly longer example would be helpful. However, I recognize I am probably not the intended audience for this paper, so it's fine to pass on this suggestion.

Minor comments:
- Typo, 'On simple'
- What does it mean "speed at which events occur"? Are we talking about time series, text streams, event time stamps, etc.?
- Typo, 'need to be done' (no period)
- What is 'the sequel'?
- Typo, 'unversal'
- Typo, 'In order to property'

File: papers/sam_morley/roughpy.md (hunk 4 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md]: It supports both, but I will clarify.

**Comment by inakleinbottle** [comment]
Thanks so much for your comments. I'm looking to fix the issues that you highlighted.

@cdlindsey I cited the Datasig website for the more complex examples, but I guess this wasn't sufficiently clear. I will adjust my wording to make this more clear.

@cliffckerr Part of the purpose of RoughPy is that non-mathematicians can make use of the complicated machinery that exists below the surface, so in some way you are exactly part of the target audience for the package. Unfortunately, rough analysis is not sufficiently well-known outside of the mathematics community that I can reference it actually defining everything. We do have a bit of space to play with in this paper, so I can certainly try to explain the connection between the example and the more involved applications. "In the sequel" is used to mean "the rest of the paper". I'm now realizing that it might not be so ubiquitous.

File: papers/sam_morley/roughpy.md (hunk 2 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md]: I've fixed this typo.

File: papers/sam_morley/roughpy.md (hunk 1 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md:43]: I've added a link to the project GitHub at the end of the introduction.

File: papers/sam_morley/roughpy.md (hunk 2 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md:38]: I've fixed this typo.

File: papers/sam_morley/roughpy.md (hunk 3 above)
Comment (inakleinbottle) [reply papers/sam_morley/roughpy.md]: The first step of the example now installs RoughPy using pip.

**Comment by inakleinbottle** [comment]
OK, I believe I have addressed all the issues raised by the reviewers, and a few additional things I noticed too. This should be my final version.

**Pull request merged**